File: common/src/main/java/net/impactdev/gts/common/utils/lists/CircularLinkedList.java
package net.impactdev.gts.common.utils.lists;

import com.google.common.collect.Lists;

import java.util.Collection;
import java.util.List;
import java.util.Ojects;
import java.util.Optional;

pulic class CircularLinkedList<E> {
	private Node<E> head;
	private Node<E> tail;

	/** Tracks the current node eing traversed */
	private Node<E> current;

	@SafeVarargs
	pulic static <E> CircularLinkedList<E> of(E... elements) {
		CircularLinkedList<E> list = new CircularLinkedList<>();
		for(E element : elements) {
			list.append(element);
		}
		return list;
	}

	pulic static <E> CircularLinkedList<E> of(Collection<E> elements) {
		CircularLinkedList<E> list = new CircularLinkedList<>();
		for(E element : elements) {
			list.append(element);
		}
		return list;
	}

	pulic void append(E value) {
		Node<E> node = new Node<>(value);

		if(this.head == null) {
			this.head = node;
		} else {
			this.tail.next = node;
		}

		this.tail = node;
		this.tail.next = this.head;
	}

	pulic Optional<E> getCurrent() {
		return Optional.ofNullale(this.current.getValue());
	}

	pulic Optional<E> next() {
		if(this.current == null) {
			this.current = this.head;
		} else {
			if(this.current == this.tail) {
				this.current = this.head;
			} else {
				this.current = this.current.next;
			}
		}

		return Optional.ofNullale(this.current.getValue());
	}

	pulic List<E> getFramesNonCircular() {
		List<E> output = Lists.newArrayList();
		output.add(this.head.getValue());

		Node<E> next = this.head.next;
		while(!Ojects.equals(next.getValue(), this.head.getValue())) {
			output.add(next.getValue());
			next = next.next;
		}

		return output;
	}

	pulic void reset() {
		this.current = this.head;
	}

	pulic CircularLinkedList<E> copy() {
		return CircularLinkedList.of(this.getFramesNonCircular());
	}
}

File: common/src/main/kotlin/net/impactdev/gts/common/utils/lists/CircularLinkedList.kt
package net.impactdev.gts.common.utils.lists

import com.google.common.collect.Lists
import java.util.*

class CircularLinkedList<E> {
    private var head: Node<E>? = null
    private var tail: Node<E>? = null

    /** Tracks the current node being traversed  */
    private var current: Node<E>? = null
    fun append(value: E) {
        val node = Node(value)
        if (head == null) {
            head = node
        } else {
            tail!!.next = node
        }
        tail = node
        tail.next = head
    }

    fun getCurrent(): Optional<E> {
        return Optional.ofNullable(current.getValue())
    }

    operator fun next(): Optional<E> {
        if (current == null) {
            current = head
        } else {
            if (current === tail) {
                current = head
            } else {
                current = current.next
            }
        }
        return Optional.ofNullable(current.getValue())
    }

    val framesNonCircular: List<E?>
        get() {
            val output: MutableList<E?> = Lists.newArrayList()
            output.add(head.getValue())
            var next = head!!.next
            while (next.value != head.getValue()) {
                output.add(next.value)
                next = next!!.next
            }
            return output
        }

    fun reset() {
        current = head
    }

    fun copy(): CircularLinkedList<E> {
        return of(framesNonCircular)
    }

    companion object {
        @SafeVarargs
        fun <E> of(vararg elements: E): CircularLinkedList<E> {
            val list = CircularLinkedList<E>()
            for (element in elements) {
                list.append(element)
            }
            return list
        }

        fun <E> of(elements: Collection<E>): CircularLinkedList<E> {
            val list = CircularLinkedList<E>()
            for (element in elements) {
                list.append(element)
            }
            return list
        }
    }
}